Engine/src/Platform/OpenGL41/OpenGL41RendererAPI.cpp

```diff
@@ -0,0 +1,147 @@
+#include "Platform/OpenGL41/OpenGL41RendererAPI.h"
+#include "Platform/OpenGL41/OpenGL41Check.h"
+#include "Platform/OpenGLCommon.h"
+
+#include <glad/gl.h>
+
+#include <string>
+
+namespace elv {
+
+void glCheckError(
+    const char* file,
+    int line,
+    const char* expression)
+{
+    GLenum errorCode;
+
+    do {
+        errorCode = glGetError();
+
+        if (errorCode != GL_NO_ERROR) {
+            std::string fileString = file;
+            std::string error = "Unknown error = " + std::to_string(errorCode);
+            std::string description = "No description";
+
+            switch (errorCode) {
+            case GL_INVALID_ENUM: {
+                error = "GL_INVALID_ENUM";
+                description = "An unacceptable value has been specified for an enumerated argument.";
+                break;
+            }
+
+            case GL_INVALID_VALUE: {
+                error = "GL_INVALID_VALUE";
+                description = "A numeric argument is out of range.";
+                break;
+            }
+
+            case GL_INVALID_OPERATION: {
+                error = "GL_INVALID_OPERATION";
+                description = "The specified operation is not allowed in the current state.";
+                break;
+            }
+
+            case GL_STACK_OVERFLOW: {
+                error = "GL_STACK_OVERFLOW";
+                description = "This command would cause a stack overflow.";
+                break;
+            }
+
+            case GL_STACK_UNDERFLOW: {
+                error = "GL_STACK_UNDERFLOW";
+                description = "This command would cause a stack underflow.";
+                break;
+            }
+
+            case GL_OUT_OF_MEMORY: {
+                error = "GL_OUT_OF_MEMORY";
+                description = "There is not enough memory left to execute the command.";
+                break;
+            }
+            }
+
+            std::string finalMessage = "An internal OpenGL call failed in " + fileString + "(" + std::to_string(line) + ").\nExpression:\n   " + expression + "\nError description:\n   " + error + "\n   " + description + "\n";
```

Comment:
Maybe const.
Also for such complex strings building `fmt::format` library can be used instead of a `+` operator for better readability.

Reply:
Good point, forgot about that one